src/methods/wallet-request-execution-permissions.test.ts

```diff
@@ -0,0 +1,201 @@
+const CHAIN_ID_MOCK = '0x1';
+const CONTEXT_MOCK = '0x123abc';
+
+const REQUEST_MOCK = {
+  params: [
+    {
+      chainId: CHAIN_ID_MOCK,
+      address: ADDRESS_MOCK,
+      signer: {
+        type: 'account',
+        data: {
+          address: ADDRESS_MOCK,
+        },
+      },
+      permission: {
+        type: 'test-permission',
+        isAdjustmentAllowed: true,
+        data: { key: 'value' },
+      },
+      rules: [
+        {
+          type: 'test-rule',
+          isAdjustmentAllowed: false,
+          data: { ruleKey: 'ruleValue' },
+        },
+      ],
+    },
+  ],
+} as unknown as JsonRpcRequest;
+
+const RESULT_MOCK: RequestExecutionPermissionsResult = [
+  {
+    chainId: CHAIN_ID_MOCK,
+    address: ADDRESS_MOCK,
+    signer: {
+      type: 'account',
+      data: { address: ADDRESS_MOCK },
+    },
+    permission: {
+      type: 'test-permission',
+      isAdjustmentAllowed: true,
+      data: { key: 'value' },
+    },
+    rules: [
+      {
+        type: 'test-rule',
+        isAdjustmentAllowed: false,
+        data: { ruleKey: 'ruleValue' },
+      },
+    ],
+    context: CONTEXT_MOCK,
+  },
+];
+
+describe('wallet_requestExecutionPermissions', () => {
+  let request: JsonRpcRequest;
+  let params: RequestExecutionPermissionsRequestParams;
+  let response: PendingJsonRpcResponse<Json>;
+  let processRequestExecutionPermissionsMock: jest.MockedFunction<ProcessRequestExecutionPermissionsHook>;
+
+  async function callMethod() {
+    return walletRequestExecutionPermissions(request, response, {
+      processRequestExecutionPermissions:
+        processRequestExecutionPermissionsMock,
+    });
+  }
+
+  beforeEach(() => {
+    jest.resetAllMocks();
+
+    request = klona(REQUEST_MOCK);
+    params = request.params as RequestExecutionPermissionsRequestParams;
+    response = {} as PendingJsonRpcResponse<Json>;
+
+    processRequestExecutionPermissionsMock = jest.fn();
+    processRequestExecutionPermissionsMock.mockResolvedValue(RESULT_MOCK);
+  });
+
+  it('calls hook', async () => {
+    await callMethod();
+    expect(processRequestExecutionPermissionsMock).toHaveBeenCalledWith(
+      params,
+      request,
+    );
+  });
+
+  it('returns result from hook', async () => {
+    await callMethod();
+    expect(response.result).toStrictEqual(RESULT_MOCK);
+  });
+
+  it('supports null rules', async () => {
+    params[0].rules = null as never;
+
+    await callMethod();
+
+    expect(processRequestExecutionPermissionsMock).toHaveBeenCalledWith(
+      params,
+      request,
+    );
+  });
+
+  it('supports optional address', async () => {
+    params[0].address = undefined as never;
+
+    await callMethod();
+
+    expect(processRequestExecutionPermissionsMock).toHaveBeenCalledWith(
+      params,
+      request,
+    );
+  });
+
+  it('throws if no hook', async () => {
+    await expect(
+      walletRequestExecutionPermissions(request, response, {}),
+    ).rejects.toMatchInlineSnapshot(
```

Comment:
Nit: Curious, why are we using a snapshot in these tests rather than `.toThrow(new Error(...))`?

Reply:
Yeah, that's a bit silly - I think it was due to the complex error messages. I've replaced it with the more succinct error reason.